### ISSUE TITLE
Support for Eloquent API Resources

### ISSUE THREAD

**Issue by aubreychiduku** [MarcinOrlowski/laravel-api-response-builder#109]
We're using [API Resources](https://laravel.com/docs/6.x/eloquent-resources#resource-responses) to filter the model attributes that are transformed in JSON.

Do you plan to support UserResources in the ResponseBuilder in addition to Collections and Arrays?

**Comment by CiiDyR** [comment]
I just came across the same issue.
For now I just use `$hidden` and `$appends` on my models and pass the model collections directly (or with a paginator). This manipulates my responses the way I want it to and I could change attribute visibility on the fly with `makeHidden` and `makeVisible`.

For paginator support is just added this to the "classes" array in the config:
`\Illuminate\Pagination\AbstractPaginator::class => [
    'key' => 'pages',
     'method' => 'toArray',
],`

But I'd love resource support, it would be a lot cleaner if you want support for pagination.

**Comment by nonovd** [comment]
There already is a key for JsonResources in the classes array of the configuration file.

```php
\Illuminate\Http\Resources\Json\JsonResource::class => [
	'key'    => 'item',
	'method' => 'toArray',
],
```

We need to set the $request argument as optionnal.

```php
class User extends JsonResource
{
    /**
     * Transform the resource into an array.
     *
     * @param  \Illuminate\Http\Request | null  $request
     * @return array
     */
    public function toArray($request=null)
    {
        return [
            'id'            => $this->hashid(),
            'name'          => $this->name,
            'email'         => $this->email,
        ];
    }
}
```

**Comment by CiiDyR** [comment]
Exactly, just wanted to comment that!

**Comment by CiiDyR** [comment]
Do you know if there is a better way to handle pagination?

**Comment by nonovd** [comment]
Sorry I don't know yet, I just started trying this package today. I'd like to know as well.

**Comment by aubreychiduku** [comment]
> There already is a key for JsonResources in the classes array of the configuration file.
> 
> ```
> \Illuminate\Http\Resources\Json\JsonResource::class => [
> 	'key'    => 'item',
> 	'method' => 'toArray',
> ],
> ```
> 
> We need to set the $request argument as optionnal.
> 
> ```
> class User extends JsonResource
> {
>     /**
>      * Transform the resource into an array.
>      *
>      * @param  \Illuminate\Http\Request | null  $request
>      * @return array
>      */
>     public function toArray($request=null)
>     {
>         return [
>             'id'            => $this->hashid(),
>             'name'          => $this->name,
>             'email'         => $this->email,
>         ];
>     }
> }
> ```

Thanks @CiiDyR, missed that in the docs

**Comment by aubreychiduku** [comment]
Thanks @nonovd, missed that in the docs

**Comment by nonovd** [comment]
You're welcome. I'm not sure it is in the docs. I think you can reopen the issue because this is still not ideal as you can see (I edited my previous answer).

**Comment by CiiDyR** [comment]
I toyed around with it a little bit, but best I was able to do was extend the ResourceCollection class.
This would however not work with `::collection()`, since that uses `AnonymousResourceCollection` to create a new ResourceCollection.

It would be awesome if the converter config could define a class to handle the conversion logic, just a simple key doesn't cut it sometimes.

**Comment by MarcinOrlowski** [comment]
Thanks for the feedback. The `key` value should be honored so this may be a bug indeed. As for JsonResource - it's already supported out of the box (btw your config example misses coma separating array entries, but I expect to be just typo in your post). As for `toArray()` signature - what if you'd add own `toArray()` to your class extending `JsonResource` and simply do `return parent->toArray(null);` there? Would that work that around?

**Comment by aubreychiduku** [comment]
> I toyed around with it a little bit, but best I was able to do was extend the ResourceCollection class.
> This would however not work with `::collection()`, since that uses `AnonymousResourceCollection` to create a new ResourceCollection.
> 
> It would be awesome if the converter config could define a class to handle the conversion logic, just a simple key doesn't cut it sometimes.

I managed to get this working by accessing the underlying mapped collection with `ModelResource::collection($modelCollection)->collection`

**Comment by nonovd** [comment]
@aubreychiduku Thanks and we get the correct json structure back with the "items" key.

**Comment by CiiDyR** [comment]
This doesn't work with pagination unfortunately. I'll stick with my workaround for now

**Comment by MarcinOrlowski** [comment]
@nonovd as for your "The problem is the response will use the "item" key instead of "items"." can you please check if it is really that entry that is being used - I mean because of how RB tries to find best matching config for the class it may be that some other class matches. I.e. If you have class A that extends B and C and then you add config for classes B and C then passing object of A (which is not explicitly specified in config) will make RB look for parent class related config and it may simply pick i.e. B instead of C because of order of elements in confg array. The simplest test would be to change all the `key` entries in your config to be unique string (i.e. `key` => `key_for_class_A`) - this will let you figure out if this is really a bug and right key is not being used or (which I think is most likely) config used is different from what you think it would be.

**Comment by nonovd** [comment]
@MarcinOrlowski Yes, config used is different from what I expect. I've already tried to change the order in the config array without success. If I use unique key entries (i.e. key => key_for_class_A), it does work as expected.

**Comment by MarcinOrlowski** [comment]
Can you post your whole `classes` part of your config?

**Comment by nonovd** [comment]
@MarcinOrlowski I just put `\Illuminate\Http\Resources\Json\ResourceCollection::class` before everything else.

I checked your Converter class and found out you just loop through the classes array so there is no reason the order shouldn't be respected.

It turns out the order is lost when your merge the config files in the Service Provider. I didn't inspect it further yet.

**Comment by MarcinOrlowski** [comment]
I rewrote config merger as it was in fact broken. It's in dev branch now.

**Comment by MarcinOrlowski** [comment]
v6.3.1  is out now. Closing as fixed. Please reopen in case of problems.

**Comment by nonovd** [comment]
Still having the issue: Order of classes config key is lost.

**To Reproduce**

Package's original classes config key:
```php
	'classes'           => [
		\Illuminate\Database\Eloquent\Model::class          => [
			'key'    => 'item',
			'method' => 'toArray',
		],
		\Illuminate\Support\Collection::class               => [
			'key'    => 'items',
			'method' => 'toArray',
		],
		\Illuminate\Database\Eloquent\Collection::class     => [
			'key'    => 'items',
			'method' => 'toArray',
		],
		\Illuminate\Http\Resources\Json\JsonResource::class => [
			'key'    => 'item',
			'method' => 'toArray',
		],
	],
```

Own classes key:
```php
	'classes'           => [
		\Illuminate\Database\Eloquent\Model::class          => [
			'key'    => 'item',
			'method' => 'toArray',
		],
		\Illuminate\Support\Collection::class               => [
			'key'    => 'items',
			'method' => 'toArray',
		],
		\Illuminate\Database\Eloquent\Collection::class     => [
			'key'    => 'items',
			'method' => 'toArray',
		],
		\Illuminate\Http\Resources\Json\ResourceCollection::class => [
			'key'    => 'items',
			'method' => 'toArray',
		],
		\Illuminate\Http\Resources\Json\JsonResource::class => [
			'key'    => 'item',
			'method' => 'toArray',
		],
	],
```

**Expected behavior**

`Config::get('response_builder.classes');` should return:
```php
[
     "Illuminate\Database\Eloquent\Model" => [
       "key" => "item",
       "method" => "toArray",
     ],
     "Illuminate\Support\Collection" => [
       "key" => "items",
       "method" => "toArray",
     ],
     "Illuminate\Database\Eloquent\Collection" => [
       "key" => "items",
       "method" => "toArray",
     ],
     "Illuminate\Http\Resources\Json\ResourceCollection" => [
       "key" => "items",
       "method" => "toArray",
     ],
     "Illuminate\Http\Resources\Json\JsonResource" => [      
       "key" => "item",
       "method" => "toArray",
     ],
   ]
```

**Actual behavior**

`Config::get('response_builder.classes');` returns:

```php
[
     "Illuminate\Database\Eloquent\Model" => [
       "key" => "item",
       "method" => "toArray",
     ],
     "Illuminate\Support\Collection" => [
       "key" => "items",
       "method" => "toArray",
     ],
     "Illuminate\Database\Eloquent\Collection" => [
       "key" => "items",
       "method" => "toArray",
     ],
     "Illuminate\Http\Resources\Json\JsonResource" => [      
       "key" => "item",
       "method" => "toArray",
     ],
     "Illuminate\Http\Resources\Json\ResourceCollection" => [
       "key" => "items",
       "method" => "toArray",
     ],
   ]
```
**Env**
 - PHP 7.3
 - Laravel 6.5.0
 - Laravel Api Response Builder 6.3.1

**Comment by nonovd** [comment]
@MarcinOrlowski

**Comment by MarcinOrlowski** [comment]
Reopening. The root is that config merger is merging your config into defaults, and this would preserve the order if you are overriding existing entry (i.e. for `Illuminate\Http\Resources\Json\JsonResource`) but if there's anything new like your `ResourceCollection` then it is appended to the result config. This is apparently bad in this particular case when order matters. Let me think how to solve this w/o killing defaults and forcing users to always have config file set up.

BTW: you do not need to call me (with @) as I will get notification about every comment posted.

**Comment by nonovd** [comment]
Thanks for your fast answer. Noted, I wasn't sure you'd get the notification since the issue was closed :)

As a quick workaround, you may add the `Illuminate\Http\Resources\Json\ResourceCollection::class` key in the original config file. I agree it needs a more robust solution. Thanks for your time.

**Comment by MarcinOrlowski** [comment]
What if defaults would NOT be used at all if you have own config set? I am trying to find out the downside of this approach and the only one is that in case of upgrade of the defaults, user won't get it automatically (which would happen with current merging approach). From the other hand, if s/he did not miss support for that class that is added to the defaults, then nobody would notice.

**Comment by MarcinOrlowski** [comment]
I just release v6.3.2 which supports `pri` key in `classes` entries. All you should need to do is to have `classes` entries in your config with just these classes you want to additionally support. This will be merged with default config as it was before, but this time `pri` is used to sort classes prior using, so all
you need to do is to ensure `pri` for that class is higher than 0:

```php
"classes" => [
     "Illuminate\Http\Resources\Json\ResourceCollection" => [
       "key" => "items",
       "method" => "toArray",
       "pri" => 10,
     ],
],
```

I will most likely add support of that class to defaults in next version anyway.

**Comment by nonovd** [comment]
This works like a charm. I messed around with it a little and we can get away with any sorting that we like. Thank you.

**Comment by MarcinOrlowski** [comment]
Great. Thanks for your feedback. Closing as fixed.

**Comment by MarcinOrlowski** [comment]
FYI I just merged #122 that delegates data conversion to external classes, so you can now easily write own converters. See branch [v6.4-dev](https://github.com/MarcinOrlowski/laravel-api-response-builder/tree/v6.4-dev) and the config file.
